File: lxe_mc/code/tracking.py
"""
tracking.py
===========

Stack-based Monte Carlo tracking for photons and leptons in a large
LXe calorimeter.

Design follows Geant4: a single LIFO particle stack, processes are
sampled at each interaction, secondaries are pushed onto the stack.
Charged particle transport is condensed-history with explicit hard
bremsstrahlung sampling.

Public API:
    simulate_event(E_MeV, position, direction, geometry, rng)
        -> list of (x, y, z, dE) energy deposits

The geometry is currently a simple infinite-LXe (no boundary check)
or a cylindrical fiducial volume.
"""

from dataclasses import dataclass, field
from typing import List, Tuple
import numpy as np

from physics import (ME_MEV, NAT_LXE, RHO_LXE, EK_XE_MEV,
                     sigma_compton, dsigma_dk_brems)
from nist_data import (sigma_compton_NIST, sigma_pair_NIST, sigma_phot_NIST,
                       sigma_total_NIST, dEdx_collision_NIST,
                       csda_range_g_per_cm2_NIST)
import sampling
import thresholds as th


# =====================================================================
# Geometry: simple LXe-everywhere (infinite) or a cylinder
# =====================================================================

@dataclass
class InfiniteLXe:
    """Infinite LXe (no boundaries)."""
    def is_inside(self, pos): return True

@dataclass
class CylinderLXe:
    """Cylinder along z, radius R, half-height H, centered at origin."""
    radius_cm: float
    half_height_cm: float
    def is_inside(self, pos):
        return (pos[0]**2 + pos[1]**2 < self.radius_cm**2
                and abs(pos[2]) < self.half_height_cm)


# =====================================================================
# Track and stack
# =====================================================================

@dataclass
class Track:
    kind: str             # 'gamma', 'electron', 'positron'
    energy: float         # MeV (kinetic for leptons; total for gamma)
    position: np.ndarray  # cm
    direction: np.ndarray # unit vector
    parent_id: int = -1
    generation: int = 0


class Stack:
    """Plain LIFO stack of Track objects."""
    def __init__(self):
        self._items: List[Track] = []
    def push(self, t: Track):
        self._items.append(t)
    def pop(self) -> Track:
        return self._items.pop()
    def empty(self) -> bool:
        return len(self._items) == 0
    def __len__(self):
        return len(self._items)


@dataclass
class Deposit:
    """An energy deposit in the LXe."""
    position: np.ndarray   # cm
    energy: float          # MeV
    source: str            # 'electron', 'positron', 'photoelectric',
                           # 'auger', 'gamma_local', 'brems_local'


# =====================================================================
# Photon transport
# =====================================================================

def transport_photon(track: Track, geom, deposits: List[Deposit],
                     stack: Stack, rng: np.random.Generator) -> None:
    """Transport one gamma until escape, absorption, or below cutoff.

    May push additional secondaries (electrons, positrons, photons) onto
    the stack. Energy deposits are appended to deposits.
    """
    pos = track.position.copy()
    direction = track.direction.copy()
    E = track.energy

    while E >= th.EGAMMA_CUT_MEV:
        # NIST cross sections (per atom of Xe in cm^2)
        sC = sigma_compton_NIST(E, per_atom=True)
        sP = sigma_pair_NIST(E, per_atom=True)
        sPh = sigma_phot_NIST(E, per_atom=True)
        s_tot = sC + sP + sPh
        Sigma_tot = NAT_LXE * s_tot

        # Sample distance to next interaction
        s = sampling.sample_distance(Sigma_tot, rng)
        pos = pos + direction * s
        if not geom.is_inside(pos):
            return  # escapes

        # Sample which process
        proc = sampling.sample_process(sC/s_tot, sP/s_tot, sPh/s_tot, rng)

        if proc == "compton":
            # sample (Egp, cos_theta), update direction, push electron
            Egp, cos_t = sampling.sample_compton(E, rng)
            phi = 2*np.pi*rng.uniform()
            n_e = sampling.compton_electron_direction(cos_t, phi, E, direction)
            T_e = E - Egp
            # push electron
            stack.push(Track(kind="electron", energy=T_e, position=pos.copy(),
                             direction=n_e, parent_id=id(track),
                             generation=track.generation + 1))
            # update photon direction: rotate by cos_t, phi about old direction
            sin_t = np.sqrt(max(0.0, 1.0 - cos_t**2))
            local = np.array([sin_t*np.cos(phi), sin_t*np.sin(phi), cos_t])
            direction = sampling._rotate_to_global(local, direction)
            E = Egp
            # continue the while loop (this same gamma keeps going)

        elif proc == "pair":
            eps = sampling.sample_pair(E, rng)
            E_pos_total = eps * E
            E_ele_total = (1.0 - eps) * E
            T_pos = E_pos_total - ME_MEV
            T_ele = E_ele_total - ME_MEV
            # Sample directions (forward-peaked)
            theta_pos = sampling.pair_polar_angle(E_pos_total, rng)
            theta_ele = sampling.pair_polar_angle(E_ele_total, rng)
            phi = 2*np.pi*rng.uniform()
            # Place leptons coplanar with incoming photon
            for sign, theta, T, kind in [(1.0, theta_pos, T_pos, "positron"),
                                          (-1.0, theta_ele, T_ele, "electron")]:
                local = np.array([np.sin(theta)*np.cos(phi + (np.pi if sign < 0 else 0)),
                                   np.sin(theta)*np.sin(phi + (np.pi if sign < 0 else 0)),
                                   np.cos(theta)])
                d_lep = sampling._rotate_to_global(local, direction)
                if T > 0:
                    stack.push(Track(kind=kind, energy=T, position=pos.copy(),
                                     direction=d_lep, parent_id=id(track),
                                     generation=track.generation + 1))
            return  # photon consumed

        elif proc == "photoelectric":
            shell, E_bind = sampling.sample_phot_shell(E, rng)
            T_e = E - E_bind
            theta_e = sampling.sample_photoelectron_angle(T_e, rng)
            phi_e = 2*np.pi*rng.uniform()
            local = np.array([np.sin(theta_e)*np.cos(phi_e),
                               np.sin(theta_e)*np.sin(phi_e),
                               np.cos(theta_e)])
            d_e = sampling._rotate_to_global(local, direction)
            if T_e > 0:
                stack.push(Track(kind="electron", energy=T_e,
                                 position=pos.copy(), direction=d_e,
                                 parent_id=id(track),
                                 generation=track.generation + 1))
            # Atomic relaxation
            if shell == "K":
                outcome, E_relax = sampling.sample_atomic_relaxation_K(rng)
                if outcome == "fluorescence":
                    # Push a new photon, isotropic
                    cos_t = -1 + 2*rng.uniform()
                    phi = 2*np.pi*rng.uniform()
                    sin_t = np.sqrt(1 - cos_t**2)
                    d_g = np.array([sin_t*np.cos(phi), sin_t*np.sin(phi), cos_t])
                    stack.push(Track(kind="gamma", energy=E_relax,
                                     position=pos.copy(), direction=d_g,
                                     parent_id=id(track),
                                     generation=track.generation + 1))
                else:
                    # Auger electron deposits locally
                    deposits.append(Deposit(pos.copy(), E_relax, "auger"))
            else:
                # L/M binding: deposit locally
                deposits.append(Deposit(pos.copy(), E_bind, "auger"))
            return  # photon consumed

    # Below threshold: deposit residual energy locally
    deposits.append(Deposit(pos, E, "gamma_local"))


# =====================================================================
# Lepton transport (electron and positron)
# =====================================================================

def transport_lepton(track: Track, geom, deposits: List[Deposit],
                     stack: Stack, rng: np.random.Generator) -> None:
    """Step the electron/positron through LXe with continuous energy loss
    plus discrete bremsstrahlung. At end of range, positron annihilates."""
    pos = track.position.copy()
    direction = track.direction.copy()
    T = track.energy

    while T >= th.T_E_CUT_MEV:
        # determine step size from residual range
        R_cm = csda_range_g_per_cm2_NIST(T) / RHO_LXE  # cm
        ds = th.F_RANGE * R_cm
        ds = max(th.DS_FLOOR_CM, min(ds, th.DS_CEIL_CM))

        # collisional energy loss (continuous)
        dEdx_col = dEdx_collision_NIST(T) * RHO_LXE   # MeV/cm
        dE_col = dEdx_col * ds
        if dE_col >= T:
            # would over-shoot; reduce step
            ds = T / dEdx_col * 0.9
            dE_col = dEdx_col * ds

        # deposit collisional energy at midpoint of step
        mid_pos = pos + direction * (ds * 0.5)
        deposits.append(Deposit(mid_pos.copy(), dE_col, track.kind))
        T -= dE_col

        # advance position
        pos = pos + direction * ds
        if not geom.is_inside(pos):
            # exit detector; deposit nothing more
            return

        if T < th.T_E_CUT_MEV:
            break

        # Sample bremsstrahlung
        if T > th.K_MIN_MEV:
            sig_b = sampling._sigma_brems_above_kmin(T, th.K_MIN_MEV)
            P_brems = NAT_LXE * sig_b * ds
            if P_brems > 0.5:
                P_brems = 0.5  # safety cap; should not happen with our step sizes
            if rng.uniform() < P_brems:
                k = sampling.sample_brems(T, th.K_MIN_MEV, rng)
                if k is not None and k < T:
                    theta_g = sampling.brems_photon_angle(T, rng)
                    phi_g = 2*np.pi*rng.uniform()
                    local = np.array([np.sin(theta_g)*np.cos(phi_g),
                                       np.sin(theta_g)*np.sin(phi_g),
                                       np.cos(theta_g)])
                    d_g = sampling._rotate_to_global(local, direction)
                    stack.push(Track(kind="gamma", energy=k,
                                     position=pos.copy(), direction=d_g,
                                     parent_id=id(track),
                                     generation=track.generation + 1))
                    T -= k

    # End-of-range: deposit remaining T
    if T > 0 and geom.is_inside(pos):
        deposits.append(Deposit(pos.copy(), T, track.kind))

    # Positron: annihilation at rest
    if track.kind == "positron":
        # two 511 keV photons back-to-back, isotropic
        cos_t = -1 + 2*rng.uniform()
        phi = 2*np.pi*rng.uniform()
        sin_t = np.sqrt(1 - cos_t**2)
        d_g = np.array([sin_t*np.cos(phi), sin_t*np.sin(phi), cos_t])
        for d in [d_g, -d_g]:
            stack.push(Track(kind="gamma", energy=ME_MEV,
                             position=pos.copy(), direction=d,
                             parent_id=id(track),
                             generation=track.generation + 1))


# =====================================================================
# Top-level event simulation
# =====================================================================

def simulate_event(E_MeV: float, position=(0.0, 0.0, 0.0),
                   direction=(0.0, 0.0, 1.0),
                   geom=None, rng=None) -> List[Deposit]:
    """Simulate one primary photon entering the calorimeter.

    Returns a list of energy Deposit objects.
    """
    if geom is None:
        geom = InfiniteLXe()
    if rng is None:
        rng = np.random.default_rng()

    deposits: List[Deposit] = []
    stack = Stack()
    stack.push(Track(kind="gamma", energy=E_MeV,
                     position=np.asarray(position, dtype=float),
                     direction=np.asarray(direction, dtype=float),
                     parent_id=-1, generation=0))

    while not stack.empty():
        t = stack.pop()
        if t.generation > th.GEN_CAP:
            continue   # safety
        if t.kind == "gamma":
            transport_photon(t, geom, deposits, stack, rng)
        else:
            transport_lepton(t, geom, deposits, stack, rng)

    return deposits


def cluster_deposits_in_z(deposits: List[Deposit],
                           dz_cm: float = th.DZ_RESOLUTION_CM
                           ) -> List[Tuple[float, float]]:
    """Group deposits into clusters whose z-extent is within dz_cm.

    Returns list of (z_centroid, total_energy). Used for SS/MS classification.
    """
    if not deposits:
        return []
    items = sorted(deposits, key=lambda d: d.position[2])
    clusters = []
    cur_zs = [items[0].position[2]]
    cur_es = [items[0].energy]
    for d in items[1:]:
        if d.position[2] - cur_zs[-1] < dz_cm:
            cur_zs.append(d.position[2])
            cur_es.append(d.energy)
        else:
            E_tot = sum(cur_es)
            z_cent = sum(z*e for z, e in zip(cur_zs, cur_es)) / E_tot
            clusters.append((z_cent, E_tot))
            cur_zs = [d.position[2]]
            cur_es = [d.energy]
    E_tot = sum(cur_es)
    z_cent = sum(z*e for z, e in zip(cur_zs, cur_es)) / E_tot
    clusters.append((z_cent, E_tot))
    return clusters


def is_single_site(deposits: List[Deposit],
                    dz_cm: float = th.DZ_RESOLUTION_CM) -> bool:
    return len(cluster_deposits_in_z(deposits, dz_cm)) == 1


if __name__ == "__main__":
    rng = np.random.default_rng(42)
    print("Simulating 100 events of 2.615 MeV gammas in infinite LXe...")
    n_ss = 0
    total_E = []
    for i in range(100):
        deposits = simulate_event(2.615, rng=rng)
        E_dep = sum(d.energy for d in deposits)
        total_E.append(E_dep)
        if is_single_site(deposits):
            n_ss += 1
    print(f"  Mean total deposit: {np.mean(total_E):.4f} MeV")
    print(f"  std total deposit: {np.std(total_E):.4f} MeV")
    print(f"  Range: {np.min(total_E):.4f} to {np.max(total_E):.4f}")
    print(f"  SS fraction (3mm in z, infinite LXe): {n_ss}/100 = {n_ss/100*100:.1f}%")

File: lxe_mc/LXeMC/src/tracking.jl
"""
Stack-based Monte Carlo transport for photons and leptons.

Transport occurs inside a `PhysicalVolume` (typically a `PCyl` filled
with an active material). Material properties (cross sections, stopping
powers) come from the `Material` attached to the volume. Boundary
checking uses the volume's `is_inside` method.

# Public API

- [`simulate_event`](@ref): full simulation (photon + electron transport).
- [`simulate_event_photon_only`](@ref): photon-only mode (electrons deposit locally).
- [`is_single_site`](@ref): convenience wrapper.
"""

using Random

const TRANSPORT_BOUNDARY_TOL_CM = 1e-5
const TRANSPORT_BOUNDARY_PUSH_CM = 1e-4


# =====================================================================
# Track and stack
# =====================================================================

"""
    Track

A particle on the transport stack.

Fields:
- `kind`: `:gamma`, `:electron`, or `:positron`
- `energy`: kinetic energy [MeV] (for leptons) or total energy (for photons)
- `position`: 3-vector [cm]
- `direction`: unit 3-vector
- `track_id`: unique integer id (assigned by event loop counter)
- `parent_id`: track_id of parent (0 for primaries)
- `generation`: cascade depth (0 for primaries)
- `interaction`: what created this track (`:primary`, `:compton`, `:photoelectric`,
  `:pair`, `:bremsstrahlung`)
- `volume`: where this track was created (`:fv`, `:active`, `:passive`)
"""
struct Track
    kind::Symbol
    energy::Float64
    position::Vector{Float64}
    direction::Vector{Float64}
    track_id::Int
    parent_id::Int
    generation::Int
    interaction::Symbol
    volume::Symbol
end


"""
    ParticleStack

LIFO stack of `Track` objects, backed by a `Vector`.
"""
mutable struct ParticleStack
    items::Vector{Track}
end
ParticleStack() = ParticleStack(Track[])
Base.push!(s::ParticleStack, t::Track) = push!(s.items, t)
Base.pop!(s::ParticleStack) = pop!(s.items)
Base.isempty(s::ParticleStack) = isempty(s.items)
Base.length(s::ParticleStack) = length(s.items)


"""
    Deposit

An energy deposit in the detector volume.

Fields:
- `position`: 3-vector [cm]
- `energy`: deposited energy [MeV]
- `source`: origin label (`:electron`, `:positron`, `:photoelectric`,
  `:pair`, `:gamma_local`)
- `interaction`: what interaction produced this deposit (`:compton`,
  `:photoelectric`, `:pair`, `:bremsstrahlung`, `:collisional`,
  `:escaped_gamma`, `:escaped_lepton`)
- `volume`: where the deposit occurred (`:fv`, `:active`, `:passive`)
"""
struct Deposit
    position::Vector{Float64}
    energy::Float64
    source::Symbol
    interaction::Symbol
    volume::Symbol
end


# =====================================================================
# Escape-point volume classification
# =====================================================================

"""
    _classify_escape_volume(fk, pos) -> Symbol

Classify the volume at position `pos` using the fast-kernel geometry.
Returns `:active` (Skin, TopActive, BarrelActive, BottomActive),
`:passive` (LXe_passive, FC_PTFE, FC_rings, or outside detector),
or `:fv` (should not happen for escapes, but included for completeness).
"""
@inline function _classify_escape_volume(fk::FastKernelGeometry, pos::Vector{Float64})::Symbol
    region = classify_fastkernel(fk, (pos[1], pos[2], pos[3]))
    region === nothing && return :passive
    rname = region.name
    rname == "FV" && return :fv
    rname == "Skin" && return :active
    rname in ("TopActive", "BarrelActive", "BottomActive") && return :active
    return :passive
end


# =====================================================================
# Photon transport (general, takes PhysicalVolume)
# =====================================================================

"""
    transport_photon!(track, vol, fk, deposits, stack, track_counter, cfg, rng)

Transport one gamma in `vol::PhysicalVolume` until escape, absorption,
or energy falls below `Egamma_cut`. Uses `is_inside(vol, pos)` for
boundary checking. On escape, records a deposit with the escaping
energy classified by `fk`.
"""
function transport_photon!(track::Track, vol::PhysicalVolume,
                           fk::FastKernelGeometry,
                           deposits::Vector{Deposit}, stack::ParticleStack,
                           track_counter::Ref{Int},
                           cfg::SimConfig, rng::AbstractRNG)
    mat = vol.material
    pos = copy(track.position)
    dir = copy(track.direction)
    E = track.energy
    tid = track.track_id
    gen = track.generation

    while E >= cfg.Egamma_cut
        sC, sP, sPh = sigma_three(mat, E)
        s_tot = sC + sP + sPh
        Σ_tot = mat.n_atom * s_tot

        s = sample_distance(Σ_tot, rng)
        pos .= pos .+ dir .* s
        if !is_inside(vol, pos)
            evol = _classify_escape_volume(fk, pos)
            push!(deposits, Deposit(copy(pos), E, :gamma, :escaped_gamma, evol))
            return
        end

        proc = sample_process(sC / s_tot, sP / s_tot, sPh / s_tot, rng)

        if proc === :compton
            Egp, cos_t = sample_compton(E, cfg, rng)
            ϕ = 2π * rand(rng)

            n_e = compton_electron_direction(cos_t, ϕ, E, dir, cfg)
            T_e = E - Egp
            track_counter[] += 1
            push!(stack, Track(:electron, T_e, copy(pos), n_e,
                               track_counter[], tid, gen + 1,
                               :compton, :fv))

            sin_t = sqrt(max(0.0, 1.0 - cos_t^2))
            local_vec = Float64[sin_t * cos(ϕ), sin_t * sin(ϕ), cos_t]
            dir = rotate_to_global(local_vec, dir)
            E = Egp

        elseif proc === :pair
            eps = sample_pair(E, cfg, rng)
            E_pos_total = eps * E
            E_ele_total = (1.0 - eps) * E
            T_pos = E_pos_total - cfg.me
            T_ele = E_ele_total - cfg.me

            θ_pos = pair_polar_angle(E_pos_total, cfg, rng)
            θ_ele = pair_polar_angle(E_ele_total, cfg, rng)
            ϕ = 2π * rand(rng)

            for (sign_val, θ, T, kind) in [(1.0, θ_pos, T_pos, :positron),
                                            (-1.0, θ_ele, T_ele, :electron)]
                ϕ_lep = ϕ + (sign_val < 0.0 ? π : 0.0)
                local_vec = Float64[sin(θ) * cos(ϕ_lep), sin(θ) * sin(ϕ_lep), cos(θ)]
                d_lep = rotate_to_global(local_vec, dir)
                if T > 0.0
                    track_counter[] += 1
                    push!(stack, Track(kind, T, copy(pos), d_lep,
                                       track_counter[], tid, gen + 1,
                                       :pair, :fv))
                end
            end
            return

        elseif proc === :photoelectric
            if E < mat.EK
                push!(deposits, Deposit(copy(pos), E, :photoelectric, :photoelectric, :fv))
                return
            end

            push!(deposits, Deposit(copy(pos), mat.EK, :photoelectric, :photoelectric, :fv))

            T_e = E - mat.EK
            if T_e > cfg.Te_cut
                θ_e = sample_photoelectron_angle(T_e, cfg, rng)
                ϕ_e = 2π * rand(rng)
                local_vec = Float64[sin(θ_e) * cos(ϕ_e), sin(θ_e) * sin(ϕ_e), cos(θ_e)]
                d_e = rotate_to_global(local_vec, dir)
                track_counter[] += 1
                push!(stack, Track(:electron, T_e, copy(pos), d_e,
                                   track_counter[], tid, gen + 1,
                                   :photoelectric, :fv))
            else
                push!(deposits, Deposit(copy(pos), T_e, :photoelectric, :photoelectric, :fv))
            end
            return
        end
    end

    push!(deposits, Deposit(pos, E, :gamma_local, :gamma_local, :fv))
end


# =====================================================================
# Lepton transport (general, takes PhysicalVolume)
# =====================================================================

"""
    transport_lepton!(track, vol, fk, deposits, stack, track_counter, cfg, rng)

Transport an electron or positron through `vol::PhysicalVolume` using
condensed-history stepping. Uses `is_inside(vol, pos)` for boundary
checking. On escape, records a deposit with the remaining energy
classified by `fk`.
"""
function transport_lepton!(track::Track, vol::PhysicalVolume,
                           fk::FastKernelGeometry,
                           deposits::Vector{Deposit}, stack::ParticleStack,
                           track_counter::Ref{Int},
                           cfg::SimConfig, rng::AbstractRNG)
    mat = vol.material
    pos = copy(track.position)
    dir = copy(track.direction)
    T = track.energy
    tid = track.track_id
    gen = track.generation

    while T >= cfg.Te_cut
        dEdx_col = dEdx_collision(mat, T) * mat.density
        ds = cfg.ds_step
        dE_col = dEdx_col * ds
        if dE_col >= T
            ds = T / dEdx_col * 0.9
            dE_col = dEdx_col * ds
        end

        mid_pos = pos .+ dir .* (ds * 0.5)
        push!(deposits, Deposit(copy(mid_pos), dE_col, track.kind, :collisional, :fv))
        T -= dE_col

        pos .= pos .+ dir .* ds
        if !is_inside(vol, pos)
            evol = _classify_escape_volume(fk, pos)
            push!(deposits, Deposit(copy(pos), T, track.kind, :escaped_lepton, evol))
            return
        end

        T < cfg.Te_cut && break

        if T > cfg.k_min
            sig_b = sigma_brems(mat, T)
            P_brems = min(mat.n_atom * sig_b * ds, 0.5)
            if rand(rng) < P_brems
                M_rej = brems_rejection_M(mat, T)
                k = sample_brems(T, cfg.k_min, mat.Z_eff, M_rej, cfg, rng)
                if k !== nothing && k < T
                    θ_g = brems_photon_angle(T, cfg, rng)
                    ϕ_g = 2π * rand(rng)
                    local_vec = Float64[sin(θ_g) * cos(ϕ_g), sin(θ_g) * sin(ϕ_g), cos(θ_g)]
                    d_g = rotate_to_global(local_vec, dir)
                    track_counter[] += 1
                    push!(stack, Track(:gamma, k, copy(pos), d_g,
                                       track_counter[], tid, gen + 1,
                                       :bremsstrahlung, :fv))
                    T -= k
                end
            end
        end
    end

    if T > 0.0 && is_inside(vol, pos)
        push!(deposits, Deposit(copy(pos), T, track.kind, :collisional, :fv))
    elseif T > 0.0
        evol = _classify_escape_volume(fk, pos)
        push!(deposits, Deposit(copy(pos), T, track.kind, :escaped_lepton, evol))
    end

    if track.kind === :positron && is_inside(vol, pos)
        cos_t = -1.0 + 2.0 * rand(rng)
        ϕ = 2π * rand(rng)
        sin_t = sqrt(1.0 - cos_t^2)
        d_g = Float64[sin_t * cos(ϕ), sin_t * sin(ϕ), cos_t]
        for d in [d_g, -d_g]
            track_counter[] += 1
            push!(stack, Track(:gamma, cfg.me, copy(pos), d,
                               track_counter[], tid, gen + 1,
                               :pair, :fv))
        end
    end
end


"""
    propagate_gamma(E_MeV, vol::PhysicalVolume, fk::FastKernelGeometry, cfg;
                    position, direction, rng) -> Vector{Deposit}

Propagate one gamma inside `vol` with full stack physics (photon +
lepton cascade). Returns all deposits, including escape deposits
classified by volume (`:fv`, `:active`, `:passive`).
"""
function propagate_gamma(E_MeV::Float64, vol::PhysicalVolume,
                         fk::FastKernelGeometry, cfg::SimConfig;
                         position::NTuple{3,Float64}=(0.0, 0.0, 0.0),
                         direction::NTuple{3,Float64}=(0.0, 0.0, 1.0),
                         rng::AbstractRNG=Random.default_rng())::Vector{Deposit}
    deposits = Deposit[]
    stack = ParticleStack()
    track_counter = Ref(0)
    track_counter[] += 1
    push!(stack, Track(:gamma, E_MeV,
                       Float64[position...],
                       Float64[direction...],
                       track_counter[], 0, 0,
                       :primary, :fv))

    while !isempty(stack)
        t = pop!(stack)
        t.generation > cfg.generation_cap && continue
        if t.kind === :gamma
            transport_photon!(t, vol, fk, deposits, stack, track_counter, cfg, rng)
        else
            transport_lepton!(t, vol, fk, deposits, stack, track_counter, cfg, rng)
        end
    end

    deposits
end







